sklearn/tree/_tree.pyx

```diff
@@ -446,7 +454,8 @@ cdef class BestFirstTreeBuilder(TreeBuilder):
 
         if not is_leaf:
             splitter.node_split(impurity, &split, &n_constant_features)
-            is_leaf = is_leaf or (split.pos >= end)
+            is_leaf = (is_leaf or split.pos >= end or
+                       split.improvement + EPSILON < min_impurity_decrease)
```

Comment:
I did this to avoid floating precision inconsistencies affecting the split... I'll explain clearly in a subsequent comment...